torch_xla/csrc/op_by_op_executor.cpp

```diff
@@ -0,0 +1,163 @@
+  ir::LoweringContext loctx("BuildNodeComputation");
+  size_t param_no = 0;
+  for (auto& operand : node->operands()) {
+    xla::Shape parameter_shape =
+        MakeShapeWithDeviceLayout(operand.shape(), device.hw_type);
+    xla::XlaOp param =
+        xla::Parameter(loctx.builder(), param_no, parameter_shape,
+                       absl::StrCat("param_", param_no));
+    ++param_no;
+    loctx.AssignOutputOp(operand, param);
+  }
+  for (auto& xla_op : loctx.LowerNode(node)) {
+    loctx.AddResult(xla_op);
+  }
+  return ConsumeValue(loctx.Build());
+}
+
+}  // namespace
+
+OpByOpExecutor::OpByOpExecutor(size_t compile_cache_size)
+    : compile_cache_(compile_cache_size) {}
+
+std::vector<xla::ComputationClient::ExecuteChainedOp> OpByOpExecutor::BuildOps(
+    tensorflow::gtl::ArraySlice<const ir::Value> roots,
+    const std::string& device) {
+  std::vector<const ir::Node*> root_nodes;
+  root_nodes.reserve(roots.size());
+  for (auto& root : roots) {
+    root_nodes.push_back(root.node.get());
+  }
+  std::vector<const ir::Node*> post_order =
+      ir::Util::ComputePostOrder(root_nodes);
+
+  std::unordered_map<const ir::Node*, size_t> node_to_index;
+  node_to_index.reserve(post_order.size());
+  for (size_t i = 0; i < post_order.size(); ++i) {
+    node_to_index[post_order[i]] = i;
+  }
+
+  Device exec_device(device);
+  std::vector<size_t> cache_keys;
+  std::unordered_map<size_t, std::vector<size_t>> compile_indices;
+  std::list<xla::Shape> compile_shapes;
+  std::vector<xla::ComputationClient::CompileInstance> compile_instances;
+  std::vector<xla::ComputationClient::ExecuteChainedOp> chained_exec_ops(
+      post_order.size());
+  for (size_t i = 0; i < post_order.size(); ++i) {
+    const ir::Node* node = post_order[i];
+    xla::ComputationClient::ExecuteChainedOp& cxop = chained_exec_ops[i];
+    const ir::ops::DeviceData* device_data =
+        dynamic_cast<const ir::ops::DeviceData*>(node);
+    if (device_data != nullptr) {
+      cxop.device_data = device_data->data();
+    } else {
+      size_t cache_key = ComputeNodeKey(node);
+      cxop.computation = compile_cache_.Get(cache_key);
+      if (cxop.computation == nullptr) {
+        XLA_COUNTER("OpByOpCompileCacheMiss", 1);
+
+        // Within a single IR graph, there can be many duplicated IR nodes, so
+        // make sure we do not issue an XLA compilation for each one of those.
+        auto& cache_key_indices = compile_indices[cache_key];
+        cache_key_indices.push_back(i);
+        if (cache_key_indices.size() == 1) {
+          cache_keys.push_back(cache_key);
+
+          xla::XlaComputation computation =
+              BuildNodeComputation(node, exec_device);
+          xla::ProgramShape program_shape =
+              ConsumeValue(computation.GetProgramShape());
+          compile_shapes.push_back(MakeShapeWithDeviceLayout(
+              program_shape.result(), exec_device.hw_type));
+          compile_instances.push_back(
```

Comment:
What's the reason for which we can't finish the compilation here, some dependency I'm not aware of or you're just preparing for multi-threading?

Reply:
MultiThreading. We already support it on the XRT side.